include/rocksdb/fs_env.h

```diff
@@ -0,0 +1,1323 @@
+
+// The FSEnv, FSSequentialFile, FSRandomAccessFile, FSWritableFile,
+// FSRandomRWFileclass, and FSDIrectory classes define the interface between
+// RocksDB and storage systems, such as Posix filesystems,
+// remote filesystems etc.
+// The interface allows for fine grained control of individual IO operations,
+// such as setting a timeout, prioritization, hints on data placement,
+// different handling based on type of IO etc.
+// This is accomplished by passing an instance of IOOptions to every
+// API call that can potentially perform IO. Additionally, each such API is
+// passed a pointer to a IODebugContext structure that can be used by the
+// storage system to include troubleshooting information. The return values
+// of the APIs is of type IOStatus, which can indicate an error code/sub-code,
+// as well as metadata about the error such as its scope and whether its
+// retryable.
+class FSEnv {
+ public:
+  struct FileAttributes {
+    // File name
+    std::string name;
+
+    // Size of file in bytes
+    uint64_t size_bytes;
+  };
+
+  // Priority of an IO request. This is a hint and does not guarantee any
+  // particular QoS.
+  // IO_LOW - Typically background reads/writes such as compaction/flush
+  // IO_HIGH - Typically user reads/synchronous WAL writes
+  enum class IOPriority : uint8_t {
+    IO_LOW,
+    IO_HIGH,
+    IO_TOTAL,
+  };
+
+  // Type of the data begin read/written. It can be passed down as a flag
+  // for the FSEnv implementation to optionally handle different types in
+  // different ways
+  enum class IOType : uint8_t {
+    kData,
+    kFilter,
+    kIndex,
+    kMetadata,
+    kWAL,
+    kManifest,
+    kLog,
+    kUnknown,
+    kInvalid,
+  };
+
+  // Per-request options that can be passed down to the FSEnv implementation.
+  // These are hints and are not necessarily guaranteed to be honored.
+  // More hints can be added here in the future to indicate things like
+  // storage media (HDD/SSD) to be used, replication level etc.
+  struct IOOptions {
+    // Timeout for the operation in milliseconds
+    std::chrono::milliseconds timeout;
+
+    // Priority - high or low
+    IOPriority prio;
+
+    // Type of data being read/written
+    IOType type;
+  };
+
+  // A structure to pass back some debugging information from the FSEnv
+  // implementation to RocksDB in case of an IO error
+  struct IODebugContext {
+    // file_path to be filled in by RocksDB in case of an error
+    std::string file_path;
+
+    // To be set by the FSEnv implementation
+    std::string msg;
+
+    IODebugContext(IOOptions& options) { opts = options; }
+
+    std::string ToString() {
+      std::ostringstream ss;
+      ss << file_path << ", ";
+      ss << opts.timeout << ", ";
+      ss << opts.prio << ", ";
+      ss << opts.type << ", ";
+      ss << msg;
+      return ss.str();
+    }
+
+   private:
+    IOOptions opts;
+  };
+
+  FSEnv() {}
+
+  virtual ~FSEnv();
+
+  static const char* Type() { return "Environment"; }
```

Comment:
This needs to be something other than "Environment", like "FSEnvironment", to prevent it from clashing with the Env value.